### PR TITLE
CAMEL-10349, javaScript setHeader conversion fault

### PR DESCRIPTION
**Pull request by Fabryprog**
https://issues.apache.org/jira/browse/CAMEL-10349

### CODE REVIEW

**Comment by Fabryprog** [comment]
I added the test file "JavaScriptListWithMapExpressionTest.java" to reproduce the BUG

File: components/camel-script/src/main/java/org/apache/camel/builder/script/ScriptBuilder.java [hunk 1]
```diff
@@ -32,6 +32,9 @@
 import javax.script.ScriptEngineManager;
 import javax.script.ScriptException;
 
+import jdk.nashorn.api.scripting.NashornScriptEngine;
+import jdk.nashorn.api.scripting.ScriptObjectMirror;
```
Comment (davsclaus) [inline components/camel-script/src/main/java/org/apache/camel/builder/script/ScriptBuilder.java]: Not sure if we can import this package in Camel code as it should only use the public api from JDK, eg that is typically java. For example not sure if OpenJDK or IBM JDKs have this jdk.nashorn package.

Can you research some more on the internet about this

Comment (Fabryprog) [inline components/camel-script/src/main/java/org/apache/camel/builder/script/ScriptBuilder.java]: Ok, i am using openJDK!

**Comment by davsclaus** [comment]
@gnodet @jbonofre @cschneider  do you guys know about the `jdk.nashorn` package in Java 8 is official and that its safe to import and use in Java code?

I cannot find this package exported in Apache Karaf in its `etc/jre.properties` file and hence think it will cause problems on OSGi.

**Comment by Fabryprog** [comment]
I beg you to give special attention to this PR, as this scripting runtime code is evaluated and it is very difficult to predict on a big project

**Comment by Fabryprog** [comment]
Do you have any news?

**Comment by davsclaus** [comment]
No the OSGi experts haven't responded.

**Comment by Fabryprog** [comment]
@davsclaus tries to find a solution otherwise I do not think you have to release camel 2:18 with this backward incompatibility!!

**Comment by oscerd** [comment]
There are always backward incompatibilities in release like this.

**Comment by Fabryprog** [comment]
certainly it is always best to reduce them :-)

**Comment by oscerd** [comment]
Anyway looking at this document

https://docs.oracle.com/javase/8/docs/technotes/guides/scripting/nashorn/toc.html

nashorn should be part of Java SE.

**Comment by Fabryprog** [comment]
@oscerd the problem is OSGi incompatibility! :-1:

**Comment by oscerd** [comment]
you're right. Missed the messages from @davsclaus.

**Comment by jbonofre** [comment]
R: @jbonofre

**Comment by Fabryprog** [comment]
@davsclaus do you decided to leave out this PR? It will not be included in 2.18?

File: components/camel-script/src/main/java/org/apache/camel/builder/script/ScriptBuilder.java [hunk 2]
```diff
@@ -481,6 +484,13 @@ protected Object runScript(ScriptEngine engine, Exchange exchange, ScriptContext
         if (result == null) {
             result = engine.get("result");
         }
+        
+        // into jdk8 a list will be mapped to ScriptObjectMirror
+        if (engine instanceof NashornScriptEngine && result != null && result instanceof ScriptObjectMirror) {
```
Comment (apupier) [inline components/camel-script/src/main/java/org/apache/camel/builder/script/ScriptBuilder.java]: you can avoid result != null check, result instanceof ScriptObjectMirror will return false if it is null

Comment (apupier) [inline components/camel-script/src/main/java/org/apache/camel/builder/script/ScriptBuilder.java]: you're welcome, unfortunately no, I was just reading the code to learn about this project so I'm not able to answer you about the OSGi compatibility, sorry.

**Comment by jbonofre** [comment]
Hi guys,
Nashorn is part of the Oracle JSE and provides `javax.script.*` AFAIR.
I have to check if it's also part of the IBM JDK.

Anyway, it's not a big deal to add in `jre.properties`. If the package is not present it won't be exported. 

Let me check to update Karaf default `jre.properties`.

**Comment by davsclaus** [comment]
@jbonofre did you fix something in Karaf?

**Comment by jbonofre** [comment]
@davsclaus not yet, I will do it later today (with the corresponding Karaf Jira).

**Comment by davsclaus** [comment]
Is there a JIRA for karaf?

**Comment by Fabryprog** [comment]
@davsclaus i am watching this issue https://issues.apache.org/jira/browse/KARAF-4603

**Comment by davsclaus** [comment]
Is there any progress on the Karaf team adding that import so it would work in the future in Karaf?

**Comment by davsclaus** [comment]
Dont think there is much progress form Karaf team. Wonder if we can do reflection / dynamic loading in camel-script?

**Comment by Fabryprog** [comment]
Hello everyone... there are any news?

**Comment by oscerd** [comment]
From the Karaf team no news.

**Comment by dmvolod** [comment]
Please pay attention, that KARAF-4603 issue has been fixed

**Comment by oscerd** [comment]
This is fixed on Karaf 4.2.0, Karaf 4.1.2 and Karaf 4.0.10

**Comment by oscerd** [comment]
I think this can be closed

**Comment by davsclaus** [comment]
Even so I think we need to make this change dynamic so it works on older karaf versions. If the Karaf team was quicker and fixed this about a year ago then we could assume that users was using a recent version etc. But we may have Karaf 4.0.8 and 4.1.1 users that will try to run Camel 2.20.0 etc and then fail.

Then they wont even be able to install camel-script due that osgi import problem. And they would need to fiddle with the jre.properties file. And if they have karaf running in production they cannot restart the JVM etc.

So it would have to be defined as optional imports in camel-script. And then some dynamic code that attempts to load those nashhorn classes and turn this on|off accordingly if its available or not.

**Comment by davsclaus** [comment]
The Karaf team have not yet fixed this properly, installing camel-script now fails with

```
karaf@root()> feature:install camel-script-javascript
Error executing command: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=camel-script-javascript; type=karaf.feature; version="[2.21.0.SNAPSHOT,2.21.0.SNAPSHOT]"; filter:="(&(osgi.identity=camel-script-javascript)(type=karaf.feature)(version>=2.21.0.SNAPSHOT)(version<=2.21.0.SNAPSHOT))" [caused by: Unable to resolve camel-script-javascript/2.21.0.SNAPSHOT: missing requirement [camel-script-javascript/2.21.0.SNAPSHOT] osgi.identity; osgi.identity=camel-script; type=karaf.feature; version="[2.21.0.SNAPSHOT,2.21.0.SNAPSHOT]" [caused by: Unable to resolve camel-script/2.21.0.SNAPSHOT: missing requirement [camel-script/2.21.0.SNAPSHOT] osgi.identity; osgi.identity=org.apache.camel.camel-script; type=osgi.bundle; version="[2.21.0.SNAPSHOT,2.21.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve org.apache.camel.camel-script/2.21.0.SNAPSHOT: missing requirement [org.apache.camel.camel-script/2.21.0.SNAPSHOT] osgi.wiring.package; filter:="(osgi.wiring.package=jdk.nashorn.api.scripting)"]]]
```

**Comment by davsclaus** [comment]
Its still a problem in Karaf and KARAF-4603 did not fix this.

You have to add the osgi import in the etc/jre.properties file to make it work.
@gnodet wonder if you could take a 2nd look at your fix and possible amend it to include the osgi package in the etc/jre.properties file as well.

**Comment by davsclaus** [comment]
The Karaf guys keeps making this wrong, so lets just give up.

**Pull request closed without merging**